core/src/main/java/com/linecorp/armeria/client/HAProxyHandler.java

```diff
@@ -0,0 +1,112 @@
+        ctx.pipeline().addBefore(ctx.name(), null, HAProxyMessageEncoder.INSTANCE);
+        super.handlerAdded(ctx);
+    }
+
+    @Override
+    public void connect(ChannelHandlerContext ctx, SocketAddress remoteAddress,
+                        SocketAddress localAddress, ChannelPromise promise) throws Exception {
+        promise.addListener(f -> {
+            if (!f.isSuccess()) {
+                return;
+            }
+            try {
+                ctx.write(createMessage(haProxyConfig, ctx.channel())).addListener(f0 -> {
+                    if (f0.isSuccess()) {
+                        ctx.pipeline().remove(HAProxyMessageEncoder.INSTANCE);
+                    } else {
+                        ctx.fireExceptionCaught(new ProxyConnectException(f.cause()));
+                        ctx.close();
+                    }
+                });
+            } catch (RuntimeException e) {
+                ctx.channel().eventLoop().execute(
+                        () -> ctx.pipeline().fireUserEventTriggered(new ProxyConnectException(e)));
+                ctx.close();
+            } finally {
+                ctx.pipeline().remove(this);
+            }
+        });
+        super.connect(ctx, remoteAddress, localAddress, promise);
+    }
+
+    private static HAProxyMessage createMessage(HAProxyConfig haProxyConfig,
+                                                Channel channel) throws ProxyConnectException {
+        final InetSocketAddress srcSocketAddress =
+                haProxyConfig.srcAddress() != null ? haProxyConfig.srcAddress()
+                                                   : (InetSocketAddress) channel.localAddress();
+        final InetSocketAddress destSockAddress = haProxyConfig.proxyAddress();
+        assert destSockAddress != null;
+
+        final InetAddress srcAddress = srcSocketAddress.getAddress();
+        final InetAddress destAddress = destSockAddress.getAddress();
+
+        if (srcAddress instanceof Inet4Address && destAddress instanceof Inet4Address) {
+            return new HAProxyMessage(V2, PROXY, TCP4,
+                                      srcAddress.getHostAddress(),
+                                      destAddress.getHostAddress(),
+                                      srcSocketAddress.getPort(), destSockAddress.getPort());
+        } else if (srcAddress instanceof Inet6Address && destAddress instanceof Inet6Address) {
+            return new HAProxyMessage(V2, PROXY, TCP4,
+                                      srcAddress.getHostAddress(),
+                                      destAddress.getHostAddress(),
+                                      srcSocketAddress.getPort(), destSockAddress.getPort());
+        } else {
+            logger.warn("Incompatible PROXY address types. srcSocketAddress: {}, destSockAddress: {}",
+                        srcAddress.getClass(), destAddress.getClass());
+            throw new IllegalArgumentException("incompatible addresses");
```

Comment:
- How about throwing a `ProxyConnectException` instead of `IAE`?
- Could you include the source and destination addresses in the exception message?

Reply:
> How about throwing a ProxyConnectException instead of IAE

This was intentional since 1) this is a private method 2) I didn't want to deal with checking if `ProxyConnectException` and wrapping in the upper level.
Let me know if you prefer otherwise though.

Reply:
I'd prefer throwing `ProxyConnectException`, so that we do not have to expose `IAE` in user's stack trace.